### ISSUE TITLE
Error: Connection attempt failed. Ensure the Spotify app is installed, and try to reconnect. Stream error. Reconnect the transport to the Spotify app. The operation couldn’t be completed. Connection refused

### ISSUE THREAD

**Issue by dhavalpkakadiya** [cjam/react-native-spotify-remote#158]
Hello @cjam

I am facing an issue when using Spotify remote on the iOS app, I’ve been trying to debug this issue for multiple days now with the same trouble.

Spotify authorization succeeds and then connection to Spotify fails with this error: **Connection attempt failed, Ensure the Spotify app is installed, and try to reconnect. Stream error. Reconnect the transport to the Spotify app. The operation couldn’t be completed. Connection refused.**

![IMG_0067](https://user-images.githubusercontent.com/72404856/133595925-10bcac52-b68e-403a-bef1-f3ecb64f3950.PNG)

Thanks, @cjam for this library!
React-Native Version: 0.64.0
 react-native-spotify-remote: 0.3.9

**Comment by cjam** [comment]
Have you tried closing spotify and re-authenticating?

**Comment by khoitruongg** [comment]
Hi, i'm facing the same issue. The authentication succeed, so if i try to close Spotify app and try again, it will not call to authenticate again and just call connect function

**Comment by dhavalpkakadiya** [comment]
> Have you tried closing Spotify and re-authenticating?

**@cjam** yes, I have check this flow multiple time

**Comment by cjam** [comment]
@MudulOzan did you ever figure this out?

**Comment by MudulOzan** [comment]
> @MudulOzan did you ever figure this out?

Kinda gave up :P focused on my react js projects had many problems with react native

**Comment by cjam** [comment]
Fair enough. Some of these issues are challenging to sort out. I wish I could get a little more support from Spotify with some of them.

**Comment by MudulOzan** [comment]
> Fair enough. Some of these issues are challenging to sort out. I wish I could get a little more support from Spotify with some of them. 

Thanks for your help. I will update this issue if I get rid of my real life issues and decide to come back at this. Thanks for your efforts, I really appreciate it.

**Comment by mickaeloak** [comment]
Facing this issue as well.

For me after a while when a song is paused, the remote automatically disconnects (it might be the expected behaviour?)
But the problem is after if I try to connect to the remote again I get that exact error message. So the user is forced to re-authenticate again and my app opens Spotify again, which is very annoying in terms of user experience.

Aren't we supposed to be able to connect again with a valid auth session?

Thanks,

**Comment by gijsfonk** [comment]
Currently experiencing this issue as well. The token should still be valid as `SpotifyAuth.getSession()` shows that the token is not expired, however, trying to (re)connect using that same token, I just get the same error as everyone else here.

Has anyone been able to solve this issue in the meantime?

Thanks

**Comment by vishalsiddhi** [comment]
@cjam I am facing the same issue. Do you have any update regarding this?
![IMG_1382](https://user-images.githubusercontent.com/7235480/200376957-e8fd3b9e-b54e-4ad8-a7f5-d711bf6c7a14.PNG)
Here is my error!!

**Comment by vishaltanna** [comment]
@cjam @mickaeloak @MudulOzan @khoitruongg Can anyone help me to resolve above error? I had tried my best but still didn't able to resolve this issue.

**Comment by MudulOzan** [comment]
@vishaltanna I found no solution to this

**Comment by cjam** [comment]
Hey, sorry for the lack of feedback from me.  I'm swamped with work so this gets pushed down the priority.

For those of you who are having an issue with this, can you please include the following information:

### Required Information

**Platform**:  
**Emulated?**:
**Os Version**: 
**React Native Version**:
**Package Version**: 

Without this, it's hard to get to the bottom of the issue.


### iOS Backgrounding / Disconnection

For those who are talking about the backgrounding on iOS, it's a known issue / limitation of the Spotify Remote SDK.  I've mentioned it to their developer advocates, but alas, it still exists.  It's because iOS will aggressively background applications that aren't using certain functionality (i.e. when you pause the music, spotify is no longer using the audio stream).  Therefore, we have to hack around it, a few options:
- Play silence (there are many silent tracks on spotify)
- reconnect the library when you are about to resume playback

**Comment by vishaltanna** [comment]
@cjam 

Platform: iOS
Emulated: iPhone SE (Real Device)
Os Version: 16.1
React Native Version:0.60.0
Package Version:0.3.10

Here is detail which you asked me to check it out further.

It will be great if you help me to sort it out this issue.

**Comment by Xal3ph** [comment]
> 

I'm trying to reconnect the library by using endSession() and authenticate() based on the AppContext provided in the example, it's not reconnecting the library when I'm about to resume playback.  What did you have in mind for this solution?

**Comment by MohammadKurjieh** [comment]
I am a bit late to the party, I managed to narrow down the issue (valid session + unable to control). Even though the issue seams random, I noticed a repeating pattern. If I have no valid session, the app opens Spotify and authenticates successfully. However, if Spotify was not already running in the background and playing a song at that time it will fail and give the error above, even-though you have a valid session. In this case, switching back to Spotify and starting the playback from there, fixes the issue when you come back to your app.

I managed to find a solution to this that has been working reliably on my test devices (needs further verification). If you set `playURI` in the authorize config to an empty string, it will resume playback of the previous song. However, you can also set that to be any song you would like. The only drawback is that this will instantly start a playback as soon as you are authenticated.

To resolve the above issue, you can play a silent music and pause it as soon as the remote is connected.

**Comment by luong-le** [comment]
I managed to fix this by endSession before authorize again.

**Comment by oguzhantasimaz** [comment]
> > Hello, I am facing the similar issue. Authentication succeed, I tried to close spotify app and authenticate and also opened spotify and then authenticate both didn't work.
> > `Couldn't authorize with or connect to Spotify Error: Spotify connection failed: could not find the Spotify app, it may need to be installed.`
> 
> It requires Spotify app is playing when you connect, you should set a default song url when set up before trying to connect

This worked thank you so much, I was going crazy!
